### PR TITLE
Add idx/QueryMatcher

### PR DESCRIPTION
**Pull request by prateek**

### CODE REVIEW

File: idx/query_matcher_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	require.True(t, idx.NewQueryMatcher(t, tq).Matches(tq))
+}
+
+func TestQueryMatcherRegexpQuery(t *testing.T) {
+	q, err := idx.NewRegexpQuery([]byte("abc"), []byte("def"))
+	require.NoError(t, err)
+	require.True(t, idx.NewQueryMatcher(t, q).Matches(q))
+}
+
+func TestQueryMatcherTermRegexpMismatch(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (prateek) [reply idx/query_matcher_test.go]: None, added the other tests

File: idx/query_matcher.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+			return q.Queries[0], true
+		}
+	}
+	return nil, false
+}
+
+func (m *queryMatcher) matches(expected, observed search.Query) bool {
+	switch q := expected.(type) {
+	case *query.TermQuery:
+		oq, ok := m.headElem(observed)
```
[the remark this answers is not in the document]
Comment (prateek) [reply idx/query_matcher.go]: To match a Conjuction/Disjunction query with a single term/regexp query within it. 

i.e. 
```
q1 := NewTermQuery("a","b")
q2, _ := NewConjunctionQuery(q1)
NewQueryMatcher(q1).Matches(q2) == true
``

File: idx/query_matcher_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"testing"
+
+	"github.com/m3db/m3ninx/idx"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestQueryMatcherTermQuery(t *testing.T) {
+	tq := idx.NewTermQuery([]byte("abc"), []byte("def"))
```
[the remark this answers is not in the document]
Comment (prateek) [reply idx/query_matcher_test.go]: sure

File: idx/query_matcher_test.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			right:    idx.NewTermQuery([]byte("abc"), []byte("def1")),
+			expected: false,
+		},
+	} {
+		require.Equal(t, tc.expected, idx.NewQueryMatcher(tc.left).Matches(tc.right))
+		require.Equal(t, tc.expected, idx.NewQueryMatcher(tc.right).Matches(tc.left))
+	}
+}
+
+func TestQueryMatcherTermMismatch(t *testing.T) {
```
Comment (jeromefroe) [inline idx/query_matcher_test.go]: nit: isn't this covered by the test above?

Comment (prateek) [reply idx/query_matcher_test.go]: Sure, will remove

**Pull request merged**